File: 2017/day15.kt
val a0: Long = 699
val b0: Long = 124
val n: Long = 2147483647

fun main() {
    var a = a0
    var b = b0
    val t: Long = 65536  // 2^16
    var count = 0
    fun runA() { a = (a * 16807) % n }
    fun runB() { b = (b * 48271) % n }
    for (i in 1..40_000_000) {
        runA()
        runB()
        if (a % t == b % t) { ++count }
    }
    println("part 1 = $count")  // 600
    a = a0
    b = b0
    count = 0
    for (i in 1..5_000_000) {
        do {
            runA()
        } while (a % 4L != 0L)
        do {
            runB()
        } while (b % 8L != 0L)
        if (a % t == b % t) { ++count }
    }
    println("part 2 = $count")  // 313
}

File: 2020-julia/day15.jl
function run(nums, count)
    prev = -1
    turn = Dict{Int,Int}()
    say = 0
    for i in 1:count
        if i <= length(nums)
            say = nums[i]
            turn[prev] = i-1
            prev = say
        else
            t = get(turn, prev, -1)
            if t == -1 
                say = 0
            else
                say = i-1-t
            end
            turn[prev] = i-1
            prev = say
        end
    end
    return say
end

nums = [0,3,1,6,7,5]
println("part 1 = ", run(nums, 2020))  # 852
println("part 2 = ", run(nums, 30_000_000))  # 6007666
